chromium_src/third_party/blink/renderer/core/frame/local_frame.cc

```diff
@@ -23,4 +27,79 @@
 
 #include "src/third_party/blink/renderer/core/frame/local_frame.cc"
 
+namespace blink {
+
+namespace {
+
+static scoped_refptr<Image> ImageFromNode(const Node& node) {
+  DCHECK(!node.GetDocument().NeedsLayoutTreeUpdate());
+  DocumentLifecycle::DisallowTransitionScope disallow_transition(
+      node.GetDocument().Lifecycle());
+
+  const LayoutObject* const layout_object = node.GetLayoutObject();
+  if (!layout_object)
+    return nullptr;
+
+  if (layout_object->IsCanvas()) {
+    return To<HTMLCanvasElement>(const_cast<Node&>(node))
+        .Snapshot(kFrontBuffer);
+  }
+
+  if (!layout_object->IsImage())
+    return nullptr;
+
+  const auto& layout_image = To<LayoutImage>(*layout_object);
+  const ImageResourceContent* const cached_image = layout_image.CachedImage();
+  if (!cached_image || cached_image->ErrorOccurred())
+    return nullptr;
+  return cached_image->GetImage();
+}
+
+}  // namespace
+
+SkBitmap LocalFrame::GetImageAtViewportPoint(const gfx::Point& viewport_point) {
+  HitTestResult result = HitTestResultForVisualViewportPos(viewport_point);
+  if (!IsA<HTMLCanvasElement>(result.InnerNodeOrImageMapImage()) &&
+      result.AbsoluteImageURL().IsEmpty()) {
+    // There isn't actually an image at these coordinates.  Might be because
+    // the window scrolled while the context menu was open or because the page
+    // changed itself between when we thought there was an image here and when
+    // we actually tried to retrieve the image.
+    //
+    // FIXME: implement a cache of the most recent HitTestResult to avoid having
+    //        to do two hit tests.
+    return {};
+  }
+
+  const scoped_refptr<Image> image =
+      ImageFromNode(*result.InnerNodeOrImageMapImage());
+  if (!image.get())
+    return {};
+
+  PaintImage paint_image = image->PaintImageForCurrentFrame();
```

Comment:
can you please add a comment that this is taken from `SystemClipboard::WriteImageWithTag` and should be updated accordingly when necessary?

Reply:
I have serious concerns here with copying code. These will not get updated and that's why we don't do this as a rule. I think if we want to implement this we should submit a patch to upstream or find some way to do it that doesn't involve copying code that 100% for sure will not get maintained as needed.

Reply:
Most of logic was copied but it's separated method. It doesn't affect upstream method.
If this method doesn't work, `TextRecognitionBrowserTest` will detect it.